sktime/annotation/ggs.py

```diff
@@ -0,0 +1,296 @@
+        for start, stop in zip(change_points[:-1], change_points[1:]):
+            segment = data[start:stop, :]
+            nrows, ncols = segment.shape
+            cov = np.cov(segment.T, bias=True)
+            log_likelihood -= self.log_likelihood(cov, nrows, ncols)
+        return log_likelihood
+
+    def add_new_change_point(self, data: np.ndarray) -> Tuple[int, float]:
+        """
+        Add break point.
+
+        Returns
+        -------
+            index: change point index
+            gained log likehood
+        """
+        # Initialize parameters
+        m, n = data.shape
+        orig_mean = np.mean(data, axis=0)
+        orig_cov = np.cov(data.T, bias=True)
+        origLL = self.log_likelihood(orig_cov, m, n)
+        totSum = m * (orig_cov + np.outer(orig_mean, orig_mean))
+        muLeft = data[0, :] / n
+        muRight = (m * orig_mean - data[0, :]) / (m - 1)
+        runSum = np.outer(data[0, :], data[0, :])
+        # Loop through all samples
+        # find point where breaking the segment would have the largest LL increase
+        minLL = origLL
+        minInd = 0
+        for i in range(2, m - 1):
+            # Update parameters
+            runSum = runSum + np.outer(data[i - 1, :], data[i - 1, :])
+            muLeft = ((i - 1) * muLeft + data[i - 1, :]) / (i)
+            muRight = ((m - i + 1) * muRight - data[i - 1, :]) / (m - i)
+            sigLeft = runSum / (i) - np.outer(muLeft, muLeft)
+            sigRight = (totSum - runSum) / (m - i) - np.outer(muRight, muRight)
+
+            # Compute Cholesky, LogDet, and Trace
+            Lleft = np.linalg.cholesky(sigLeft + float(self.lamb) * np.identity(n) / i)
+            Lright = np.linalg.cholesky(
+                sigRight + float(self.lamb) * np.identity(n) / (m - i)
+            )
+            llLeft = 2 * sum(map(math.log, np.diag(Lleft)))
+            llRight = 2 * sum(map(math.log, np.diag(Lright)))
+            (trLeft, trRight) = (0, 0)
+            if self.lamb > 0:
+                trLeft = math.pow(np.linalg.norm(np.linalg.inv(Lleft)), 2)
+                trRight = math.pow(np.linalg.norm(np.linalg.inv(Lright)), 2)
+            LL = (
+                i * llLeft
+                - float(self.lamb) * trLeft
+                + (m - i) * llRight
+                - float(self.lamb) * trRight
+            )
+            # Keep track of the best point so far
+            if LL < minLL:
+                minLL = LL
+                minInd = i
+        # Return break, increase in LL
+        return minInd, minLL - origLL
+
+    def adjust_change_points(
+        self, data: np.ndarray, change_points: List[int], newInd
+    ) -> List[int]:
+        """
+        Adjust change points.
+
+        Args
+        ----
+            data: time series data
+            change_points: change points indexes
+            newInd:
+
+        Returns
+        -------
+            change_points:
+        """
+        bp = change_points[:]
+        random.seed(0)
```

Comment:
Another interesting way to approach this can be to randomly generate the seed, but put it in the logger, so that if weird stuff comes up it is reproducible. (and can sometimes help find more edge case stuff than just using a fixed seed while in debug mode).

Reply:
I pulled `random_state` into intialization